cmd/publishing-bot/config/rules.go

```diff
@@ -0,0 +1,85 @@
+package config
+
+import (
+	"crypto/tls"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	yaml "gopkg.in/yaml.v2"
+)
+
+// Coordinate of a piece of code
+type Coordinate struct {
+	Repository string `yaml:"repository"`
+	Branch     string `yaml:"branch"`
+	// Dir from repo root
+	Dir string `yaml:"dir,omitempty"`
+}
+
+func (c Coordinate) String() string {
+	return fmt.Sprintf("[repository %s, branch %s, subdir %s]", c.Repository, c.Branch, c.Dir)
+}
+
+type BranchRule struct {
+	Source      Coordinate `yaml:"source"`
+	Destination Coordinate `yaml:"destination"`
+	// k8s.io/* repos the dst dependes on
+	Dependencies []Coordinate `yaml:"dependencies,omitempty"`
+}
+
+// a collection of publishing rules for a single destination repo
+type RepositoryRule struct {
+	DestinationRepository string       `yaml:"destination"`
+	Branches              []BranchRule `yaml:"branches"`
+	// if empty (e.g., for client-go), publisher will use its default publish script
+	PublishScript string `yaml:"publish-script"`
+	// not updated when true
+	Skip bool `yaml:"skipped,omitempty"`
+}
+
+type RepositoryRules struct {
+	SkippedSourceBranches []string         `yaml:"skip-source-branches"`
+	Rules                 []RepositoryRule `yaml:"rules"`
+}
+
+// LoadRules loads the repository rules either from the remote HTTP location or
+// a local file path.
+func LoadRules(config *Config) (*RepositoryRules, error) {
+	var (
+		content []byte
+		err     error
+	)
+	if strings.HasPrefix(config.RulesFile, "http://") ||
+		strings.HasPrefix(config.RulesFile, "https://") {
+		content, err = fetchFromHTTP(config.RulesFile)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		content, err = ioutil.ReadFile(config.RulesFile)
+		if err != nil {
+			return nil, err
+		}
+	}
+	var rules RepositoryRules
+	if err = yaml.Unmarshal(content, &rules); err != nil {
+		return nil, err
+	}
+	return &rules, nil
+}
+
+// fetchFromHTTP fetches the rules from http(s)://
+func fetchFromHTTP(url string) ([]byte, error) {
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
```

Comment:
Don't like InsecureSkipVerify. If we want that, this must be a command line flag.

Reply:
i think i will remove this for now... we can have init container that will fetch the rules via curl if we need? or just use secret

Reply:
Yes, init container can do that. The only advantage: we could fetch the rules on every run (the pod keep running right now, no cronjobs).

Reply:
@sttts also I think we should have option to get the rules from the repo we fetched :-) like it should have some standard path or an env var like:

REPO_RULES_PATH="./publish-rules.yaml"

Reply:
--source-rules-file hack/publish-rules.yaml